beacon-chain/db/kv/state.go

```diff
@@ -99,30 +114,112 @@ func (s *Store) SaveStates(ctx context.Context, states []iface.ReadOnlyBeaconSta
 		return errors.New("nil state")
 	}
 	multipleEncs := make([][]byte, len(states))
+	validatorsEntries := make(map[string][]byte)              // It's a map to make sure that you store only new validator entries.
+	validatorKeys := make([][]byte, len(states))              // For every state, this stores a compressed list of validator keys.
+	realValidators := make([][]*ethpb.Validator, len(states)) // It's temporary structure to restore state in memory after persisting it.
 	for i, st := range states {
 		pbState, err := v1.ProtobufBeaconState(st.InnerStateUnsafe())
 		if err != nil {
 			return err
 		}
+		if featureconfig.Get().EnableHistoricalSpaceRepresentation {
+			// store the real validators to restore the in memory state later.
+			realValidators[i] = pbState.Validators
+
+			// yank out the validators and store them in separate table to save space.
+			var hashes []byte
+			for _, val := range pbState.Validators {
+				valBytes, encodeErr := encode(ctx, val)
```

Comment:
instead of marshalling a validator and hashing it, why not simply take the root of the validator object ? that is more straightforward and is consistent with how we represent validators across prysm.
```go
rt,err := val.HashTreeRoot()
```

Reply:
+1